sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -350,20 +350,24 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
           throw new IOException(
             s"Error reading delta file $fileToRead of $this: key size cannot be $keySize")
         } else {
-          val keyRowBuffer = new Array[Byte](keySize)
+          // If key size in an existing file is not a multiple of 8, round it to multiple of 8
+          val keyAllocationSize = ((keySize + 7) / 8) * 8
+          val keyRowBuffer = new Array[Byte](keyAllocationSize)
```

Comment:
@cloud-fan @zsxwing @tdas @kunalkhamar 
I misunderstood. To store a state, `HDFSBackedStateStoreProvider` is used.

I added [a new test suite](https://github.com/apache/spark/pull/18503/files#diff-b601a70172e88e90cea0020b978d0b5bR482) to check `HDFSBackedStateStoreProvider` for storing and restoring state, as @kunalkhamar suggested [here](https://github.com/apache/spark/pull/18503#discussion_r126028838).

Do you think it makes sense?